lib/Vimscript/Builtins.idr

```diff
@@ -91,3 +103,163 @@ strpart' str start = unsafePerformIO
 ||| ```
 expand : String -> VIM_IO String
 expand str = builtin "expand" (String -> VIM_IO String) str 
+
+cursor : (line : Int) -> (col : Int) -> VIM_IO ()
+cursor line col = builtin "cursor" (Int -> Int -> VIM_IO ()) line col
+
+getcwd : VIM_IO String
+getcwd = builtin "getcwd" (VIM_IO String)
+
+||| Execute a shell command and discard the output.
+system' : String -> VIM_IO ()
+system' = builtin "system" (String -> VIM_IO ())
+
+-- %inline
+-- getpos : (mark : String) -> VIM_IO ()
+-- getpos = builtin "line" (String -> VIM_IO Int)
+
+public export
+Regex : Type
+Regex = String
+
+public export
+Filename : Type
+Filename = String
+
+public export
+FlagSet : Type
+FlagSet = String
+
+public export
+Range : Type
+Range = String
+
+entireFile : Range
+entireFile = "%"
+
+currentLine : Range
+currentLine = "."
+
+lineNum : Int -> Range
+lineNum = show
+
+lastLine : Range
+lastLine = "$"
+
+noFlags : FlagSet
+noFlags = ""
+
+public export
+data Line = Last
+
+public export
+data SubRepeatFlag = UseLastSearch | GlobalReplace
+
+public export
+data Ed 
+  = Delete
+  | Write (Maybe Filename)
+  | Sub Range Regex String FlagSet
+  | SubRepeat (List SubRepeatFlag)
+  | Global Regex Ed
+  | VGlobal Regex Ed
+  | Move (Maybe Range) Line
+  | Undo
+
+ppSubRepeatFlag : SubRepeatFlag -> String
+ppSubRepeatFlag UseLastSearch = "r"
+ppSubRepeatFlag GlobalReplace = "g"
+
+ppLine : Line -> String
+ppLine Last = "$"
+
+ppRange : Maybe Range -> String
+ppRange Nothing = ""
+ppRange (Just r) = r
+
+ppEd : Ed -> String
+ppEd Delete = "d"
+ppEd (SubRepeat fs) = "s" ++ (concat (map ppSubRepeatFlag fs))
+ppEd (Sub r re s f) = r ++ "s/" ++ re ++ "/" ++ s ++ "/" ++ f
+ppEd (Write f)
+  = case f of 
+         Nothing => "w"
+         Just x => "w " ++ x
+ppEd (Global re a) = "g/" ++ re ++ "/" ++ ppEd a
+ppEd (VGlobal re a) = "v/" ++ re ++ "/" ++ ppEd a
+ppEd (Move r l) = ppRange r ++ "m" ++ ppLine l
+
+exec : Ed -> VIM_IO ()
+exec = execute . ppEd
+
+sub' : Range -> Regex -> String -> VIM_IO ()
+sub' r re s = exec (Sub r re s noFlags)
+
+-- TODO escaping
+global : Regex -> Ed -> VIM_IO ()
+global re a = exec (Global re a)
+
+vglobal : Regex -> Ed -> VIM_IO ()
+vglobal re a = exec (VGlobal re a)
+
+v : Regex -> Ed -> VIM_IO ()
+v = vglobal
+
+g : Regex -> Ed -> VIM_IO ()
+g = global
+
+d : Ed
+d = Delete
+
+-- m$
+moveEnd : Ed
+moveEnd = Move Nothing Last
+
+check : List (Lazy Bool) -> VIM_IO ()
+check x = echo (if and x then "All tests passed" else "Some tests failed")
```

Comment:
Is this just for temporary testing?

Reply:
Yeah :) There'll quite a bit of this kind of code I'll have to be careful about.

Reply:
Fixed in https://github.com/owickstrom/idris-vimscript/pull/1/commits/243984ff026a9a81b90d5f013521921dec100d1f